### PR TITLE
Add a new library do deal with errors

### PR DESCRIPTION
**Pull request by yloiseau**
This introduce some objects, augmentations and functions to deal with
errors (null, exceptions, ...) functional style.

- `Result` type similar to Haskell `Either` or Rust `Result`
- augmentation on java `Optional`
- decorators to adapt function to return or deal with arguments of these
  types

Signed-off-by: Yannick Loiseau <yannick.loiseau@univ-bpclermont.fr>

### CODE REVIEW

File: src/main/golo/errors.golo [hunk 1]
```diff
@@ -0,0 +1,254 @@
+----
```
Comment (jponge) [inline src/main/golo/errors.golo]: Lacks a license header

File: src/main/golo/errors.golo [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+Useful augmentations, decorators and function to deal with errors in Golo.
+----
+module gololang.error.Errors
```
Comment (jponge) [inline src/main/golo/errors.golo]: How about `gololang.Errors`?

File: src/main/java/gololang/error/Result.java [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+
+/**
+ * A container object which represent the result of a maybe failing operation.
+ *
+ * <p>This object is used when chaining computations (e.g. using map-filter operations) that can
+ * produce an error. Instead of raising an exception, the operation can use this object to
+ * encapsulate the result. This is similar to {@code Optional}, but also encapsulate the type of
+ * error in the form of a {@code Throwable} instance that can be raised later.
+ *
+ * <p>This is similar to the {@code Either} or {@code Result} type in other functional languages (e.g.
```
Comment (jponge) [inline src/main/java/gololang/error/Result.java]: Great to have references :+1:

**Comment by jponge** [comment]
Interesting stuff

File: src/main/golo/errors.golo (hunk 2 above)
Comment (yloiseau) [inline src/main/golo/errors.golo]: I just wanted it to be in the same namespace as Error, but it's open.

**Comment by k33g** [comment]
I :heart: that

**Comment by k33g** [comment]
... and I'm ready to change all my code [https://github.com/ant-colony/lasius/blob/master/imports/sensors/temperature.ability.golo#L36](https://github.com/ant-colony/lasius/blob/master/imports/sensors/temperature.ability.golo#L36)

**Comment by k33g** [comment]
@jponge it would be nice if you merge it for the coming release :cat: (<- Puss in Boots' eyes)
it's so useful [https://github.com/TypeUnsafe/golo-f/blob/master/04.yloiseau.result.decorator.golo](https://github.com/TypeUnsafe/golo-f/blob/master/04.yloiseau.result.decorator.golo)

**Comment by yloiseau** [comment]
Just doing some last minute improvements :smile:

**Comment by yloiseau** [comment]
Ok for me, unless you see other interesting use cases

**Comment by yloiseau** [comment]
this last commit introduce `either`, @k33g will love :smile:

**Comment by yloiseau** [comment]
@k33g your code can now be: 
```golo
  toInt("hello"): either(
    |value| -> println("value: " + value),
    |error| -> println("-> error: " + error))
```

**Comment by k33g** [comment]
I dared not ask 🐼

**Comment by yloiseau** [comment]
Ooops nop, I forgot `reduce` on `Optional`

**Comment by jponge** [comment]
@yloiseau once you have it ping me for a last check :smiley:

**Comment by yloiseau** [comment]
@jponge I'll also add some documentation. What do you prefer: a specific section on dealing with errors or a subsection of "20. Misc. modules"?

**Comment by jponge** [comment]
Since it's a module it would better fit in _Misc. modules_ I think. It's a nice way to deal with errors, but not the only one, hence why.

**Comment by yloiseau** [comment]
@jpong ready for final review

**Comment by jponge** [comment]
:+1: great work!

**Comment by yloiseau** [comment]
Thanks :blush:

**Comment by k33g** [comment]
yes!!! A good opportunity for me to write an article on my first steps with
functional programming :)

**Comment by yloiseau** [comment]
@k33g :+1:

**Pull request merged**